src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalSessionSocket.java

```diff
@@ -309,15 +326,42 @@ public void dispatchInput(int inputSequence,
       default:
          break;
       }
-      
    }
    
    /**
     * Send output to the terminal emulator.
     * @param output text to send to the terminal
+    * @param detectLocalEcho local-echo detection
     */
-   public void dispatchOutput(String output)
+   public void dispatchOutput(String output, boolean detectLocalEcho)
    {
+      if (detectLocalEcho && !localEcho_.isEmpty())
+      {
+         String lastOutput = "";
+         while (!localEcho_.isEmpty() && lastOutput.length() < output.length())
```

Comment:
I think we could get the 95% case smooth by making three small changes to your PR:

1. Handle the case where the output is a superset rather than a subset of the local buffer (as discussed above)

2. Have an option to disable local echo entirely -- this offers an escape hatch for the small set of users for whom it's likely to be incorrect (such as people who use vi mode in readline)

3. Only do local echo if the cursor is at the EOL (I think your suggestion there is great, presuming it's easy and reliable to detect this state)